File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    void backtrack(vector<int> nums, int start, vector<int>& curr, vector<vector<int>>& itog) {
        itog.push_back(curr);

        for (int i = start; i < nums.size(); i++) {
            curr.push_back(nums[i]);
            backtrack(nums, i + 1, curr, itog);
            curr.pop_back();
        }
    }

    // [] [1] [2] [3] [1,2] [1,3] [2,3] [1,2,3]
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> itog;
        vector<int> curr;
        backtrack(nums, 0, curr, itog);
        return itog;
    }
};

File: 0078-subsets/0078-subsets.java
class Solution {
    public void backtrack_78(List<List<Integer>> itog, List<Integer> curr, int[] nums, int start){
        itog.add(new ArrayList<>(curr));

        for(int i = start; i < nums.length; i++){
            curr.add(nums[i]);

            backtrack_78(itog, curr, nums, i + 1);

            curr.remove(curr.size() - 1);
        }
    }

    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> itog = new ArrayList<>();
        List<Integer> curr = new ArrayList<>();
        backtrack_78(itog, curr, nums, 0);
        return itog;
    }
}